infra/token-transfer-server/src/tasks/transfer.js

```diff
@@ -0,0 +1,82 @@
+const moment = require('moment')
+
+const {
+  largeTransferThreshold,
+  largeTransferDelayMinutes
+} = require('../config')
+const { Transfer, TransferTask, Sequelize } = require('../models')
+const { executeTransfer } = require('../lib/transfer')
+const logger = require('../logger')
+const enums = require('../enums')
+
+const executeTransfers = async () => {
+  logger.info('Running execute transfers job...')
+
+  const outstandingTasks = await TransferTask.findAll({
```

Comment:
If we really want to be paranoid about race conditions, we should hold a lock on the TransferTask table, do the required validation, insert a new row into TransferTable and the finally release the lock... What are the chances that with Heroku 2 instances of the server get started at the same time?

Reply:
Thinking a bit more about all this, it might be simpler to take a lock on each Transfer row before starting its processing. This way, no need for a TransferTask table at all. And it may be less prone to situations where there is TransferTask row in a dangling state because the app got terminated (for a code push or other operational reason) in the middle of a task being executed...

Something like this:
```
 - get all Transfer rows ready for processing
 - for each row, do:
     - In a transaction statement:
         - Lock the row (using SELECT FOR UPDATE)
         - Make sure the row is in status Enqueue, if not error out
         - Update row status to Processing
         - Commit transaction
     - Execute the transfer
     - Update row status to Processed (or Failed)
````

Reply:
Good idea on the table wide lock. That'd only happen if we scaled up the dynos, but can never be too careful!

The row level locking was the other way I thought about doing this. I think the disadvantages are:

1. Doesn't actually prevent two transfer tasks running side by side and processing different transactions at the same time and causing nonce problems.
2. This extra table records how long it takes to run the job and what transactions ran with which job which may be useful.

I didn't fully flesh out the idea though so might be worth revisiting. Any thoughts on 1?

Reply:
Oh indeed, excellent point about the nonce... Yeah I think that's a big disadvantage of this solution. Darn blockchain always making things more complicated ;)
A TransferTask table is then probably the best way to go. Can't think of anything really better...

We'll have to be a bit careful about not deploying while there are payments being processed.
One potential way to help with that would be to have a way to indicate to the cron scheduler to not schedule any new task so that we can deploy new code. Maybe scheduler could check a value in the DB that indicates if payments are enabled or not before it executes any transfer. Maybe like a "config" table where we can store a flag. And that we would set manually via a DB query before a deployment. But perhaps not worth over optimizing. Let's see if we run into that problem before implementing something like this.